### DISCUSSION TITLE
add `/collections/{collection_id}/items/{item_id}/assets/{asset_id}` endpoints

### DISCUSSION THREAD

**Discussion by vincentsarago** [stac-utils/titiler-pgstac#133] [Show and tell]
Right now we use query parameter to pass the `asset` we want to visualize
```
/collections/{collection_id}/items/{item_id}/preview.png?assets={asset_id}&asset_bidx={asset_id}|1,2,3
```

but in theory we could do something like 

```
/collections/{collection_id}/items/{item_id}/assets/{asset_id}/preview.png?&asset_bidx={asset_id}|1,2,3
/collections/{collection_id}/items/{item_id}/assets/{asset_id}/preview.png?bidx=1,2,3
```

- It kinda make sense to have the `/assets/{asset_id}` suffix for one asset but what if we want to combine more than one? 

```
/collections/{collection_id}/items/{item_id}/assets/B01,B02,B03/preview.png
```

I'm not sure about using `,` as assets delimiter. what are other options?

- what about expression?

This is tricky but we could keep the same structure and add an `expression=True` parameter

```
/collections/{collection_id}/items/{item_id}/assets/B01+B02/preview.png?expression=True
```

we could also do something like:

```
/collections/{collection_id}/items/{item_id}/expression/B01+B02/preview.png
```

**Comment by drwelby** [comment]
`/collections/{collection_id}/items/{item_id}/assets/B01,B02,B03/preview.png` feels like it's overloading REST concepts.

`/collections/{collection_id}/items/{item_id}/assets/{asset_id}/preview.png?bidx=1,2,3` makes sense for previewing one asset.

What would follow for me would be:

`/collections/{collection_id}/items/{item_id}/assets/preview.png?asset_bidx={asset_id},{asset_id},{asset_id}`

**Comment by vincentsarago** [comment]
🙏 thanks @drwelby 

> /collections/{collection_id}/items/{item_id}/assets/preview.png?asset_bidx={asset_id},{asset_id},{asset_id}

We tend to use `assets={asset_id}&assets={asset_id}&assets={asset_id}` instead of `,` delimited values for query-string, which is why I though that using `/assets/{asset_id},{asset_id},{asset_id}/preview.png` notation.

We also have to keep in mind that:
- users might want to mix assets and select specific bands from assets 
- assets can be one bands
- expression can be for one asset or multiple assets (and we can use the `asset_as_band=True` option)

**Comment by vincentsarago** [comment]
😅 just realizing that we can already do most of this 

```python
from starlette.requests import Request
from titiler.pgstac.dependencies import get_stac_item
from titiler.core.factory import TilerFactory
from typing_extensions import Annotated
from fastapi import FastAPI

def ItemPathParams(
    request: Request,
    collection_id: Annotated[str, Path(description="STAC Collection Identifier")],
    item_id: Annotated[str, Path(description="STAC Item Identifier")],
    asset_id: Annotated[str, Path(description="STAC Asset Identifier")],
) -> str:
    """STAC Item dependency."""
    item = get_stac_item(request.app.state.dbpool, collection_id, item_id)
    asset_info = item.assets[asset]
    return asset_info.get_absolute_href() or asset_info.href,

app = FastAPI()
asset = TilerFactory(
    path_dependency=ItemPathParams,
    router_prefix="/collections/{collection_id}/items/{item_id}/assets/{asset_id}",
)
app.include_router(
    asset.router, tags=["STAC Asset"], prefix="/collections/{collection_id}/items/{item_id}/assets/{asset_id}"
)
```

☝️ will not allow mixing assets but for this you'll want to go the `/item` way 🙏
